Add WAF Select and design infrastructure (#1579)
hashicorp/web-unified-docs: content/well-architected-framework/docs/docs/optimize-systems/select-design/compute.mdx

- **Stress tests:** Simulate extreme, unnatural workloads to see how your infrastructure performs under high resource utilization. Stress tests are useful to determine how your infrastructure handles sudden spikes in traffic, and help identify how your infrastructure may break under extreme conditions.

How you run load and stress tests depends on your infrastructure, the applications running on it, and your users' behavior. For example, if your infrastructure runs a web application where users make many lightweight requests, your tests could simulate sending as many of these requests as possible across different endpoints. If your infrastructure runs a service that processes users uploads, your tests could upload many extremely large files all at once.

As you run your load and stress tests, monitor metrics such as CPU usage, memory usage, response latency, and application error rates. These metrics give important insight into how your application responds to heavy traffic.

> Terraform can create test environments that mirror your production environments with infrastructure as code (IaC). This ensures consistency between your testing and production environments, and helps reduce cost by allowing you to run your test environment only when you need it. 

Reviewer: ```suggestion
Terraform can create test environments that mirror your production environments with infrastructure as code (IaC). Mirror environments ensure consistency between your testing and production environments, and helps reduce cost by allowing you to run your test environment only when you need it. 
```

Reviewer: Not sure I 100% like how I worded this, but it needs to not be `this`

When you use Terraform to help test capacity, you gain the following benefits:

- **Reproducibility:** Your test environment configuration is version-controlled and can be recreated identically whenever needed, eliminating configuration drift between tests.
- **Cost efficiency:** Spin up test infrastructure only when running tests, then destroy it immediately after.
- **Parallel testing:** Create multiple isolated test environments simultaneously to test different configurations or scenarios without interference.
- **Reusable modules:** Use the same Terraform modules for both test and production environments, ensuring your tests accurately reflect your production code.

HashiCorp resources:
- [What is Infrastructure as Code with Terraform?](/terraform/tutorials/aws-get-started/infrastructure-as-code)
- [Running Terraform in automation](/terraform/tutorials/automation/automate-terraform)
- Create [development environments with Vagrant](/well-architected-framework/define-and-automate-processes/define/development-environment)
- [Monitor network traffic with Consul](/well-architected-framework/optimize-systems/monitor-system-health/monitor-network)

> - Learn more about tools to run load and stress tests:

Reviewer: ```suggestion

External resources:

- Learn more about tools to run load and stress tests:
```

As the workload on your infrastructure grows, know how and when to react to the scaling needs. When you understand how workload profiles grow, you can identify when to take action. With strong monitoring and alerting, you can respond to these needs before they become a problem for your users.

- **Monitoring and alerting:** [Monitoring](/well-architected-framework/optimize-systems/monitor-system-health) key metrics of your infrastructure, such as CPU, memory, and disk IOPS, lets you understand when you are approaching those bottlenecks you identified during load and stress testing. With early alerts, you can make the decision to add more resources to your infrastructure to keep the applications that depend on it running without error.
- **Continuous validation:** Along with monitoring and alerting, you can routinely evaluate workload patterns to decide when to make changes to your infrastructure architecture. Continuous validation is the practice of routinely performing checks against your infrastructure to ensure they meet certain requirements. For example, you can write a check to get the expiration date of a TLS certificate and alert you 30 days before it expires.

> - **Health checks:** Custom logic that runs periodically to check that a service is performing as expected. Often this logic lives in the application itself, and can be invoked by infrastructure such as a load balancer. This lets the infrastructure know if the application is in a state to accept traffic.

Reviewer: ```suggestion
- **Health checks:** Custom logic that runs periodically to check that a service is performing as expected. Often this logic lives in the application itself, and can be invoked by infrastructure such as a load balancer. The logic lets the infrastructure know if the application is in a state to accept traffic.
```
